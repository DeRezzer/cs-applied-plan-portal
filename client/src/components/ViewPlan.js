import React, {useState, useEffect} from "react";
import PlanTable from "./PlanTable";
import PlanMetadata from "./PlanMetadata";
import PlanComments from "./PlanComments";
import {useParams, withRouter} from 'react-router-dom';

function ViewPlan(props) {
  const [studentName, setStudentName] = useState("");
  const [userId, setUserId] = useState(null);
  const [planName, setPlanName] = useState("");
  const [status, setStatus] = useState(-1);
  const [courses, setCourses] = useState(
    [[], [{
      courseId: 0,
      credits: null,
      courseName: "",
      courseCode: "",
      prerequisites: ""
    }], []]
  );
  const [comments, setComments] = useState(
    [{
      commentId: 0,
      userId: 0,
      time: null,
      text: ""
    }]
  );
  const {planId} = useParams();

  useEffect(() => {
    async function fetchData(planId) {
      try {

        let userId = 0;
        const server = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;
        let url = `http://${server}/plan/${planId}`;
        let obj = [];

        try {
          // get plan data
          let response = await fetch(url);
          if (response.ok) {
            // get data from the response
            obj = await response.json();
            userId = obj[0][0].studentId;
            setCourses(obj);
            setPlanName(obj[0][0].planName);
            setUserId(obj[0][0].studentId);
            setStatus(parseInt(obj[0][0].status));
          } else {
            // we got a bad status code. send to 404 page
            const { history } = props;
            // if(history) history.push('/404');
            return;
          }
        } catch (err) {
          // send to 500 page if a server error happens while fetching plan
          console.log("An internal server error occurred. Please try again later.");
          const { history } = props;
          // if(history) history.push('/500');
          return;
        }

        // get user name
        url = `http://${server}/user/${userId}`;
        let response = await fetch(url);
        if (response.ok) {
          // get data from the response
          obj = await response.json();
          setStudentName(obj.firstName + " " + obj.lastName);
        }

        // get plan comments
        url = `http://${server}/plan/${planId}/comment`;
        response = await fetch(url);
        if (response.ok) {
          // get data from the response
          obj = await response.json();
          setComments(obj);
        }

      } catch (err) {
        // this is a server error
        console.log("An internal server error occurred. Please try again later.");
      }
    }
    console.log(planId);
    fetchData(planId);
  }, [planId]);

  return (
    <div className="view-plan">
      <PlanMetadata studentName={studentName} userId={userId}
        planName={planName} status={status} />
      <PlanTable courses={courses} />
      <PlanComments comments={comments} />
    </div>
  );
}
export default withRouter(ViewPlan);