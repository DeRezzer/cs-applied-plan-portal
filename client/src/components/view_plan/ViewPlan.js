/** @jsx jsx */

import {useState, useEffect} from "react";
import {css, jsx} from "@emotion/core";
import NavBar from "../Navbar";
import PlanTable from "./PlanTable";
import CreateReview from "./CreateReview";
import PlanMetadata from "./PlanMetadata";
import ActivityFeed from "./ActivityFeed";
import {useParams, withRouter} from "react-router-dom";
import PropTypes from "prop-types";
import BounceLoader  from "react-spinners/BounceLoader";

function ViewPlan(props) {

  const [currentUserDev] = useState(6); // Development: Selecting the current user
  const [currentUser, setCurrentUser] = useState(
    {
      id: 0,
      role: 0,
      firstName: "",
      lastName: ""
    }
  );
  const [loading, setLoading] = useState(true);
  const [studentName, setStudentName] = useState("");
  const [userId, setUserId] = useState(null);
  const [planName, setPlanName] = useState("");
  const [status, setStatus] = useState(-1);
  const [activity, setActivity] = useState([]);
  const [courses, setCourses] = useState(
    [[], [{
      courseId: 0,
      credits: null,
      courseName: "",
      courseCode: "",
      prerequisites: ""
    }], []]
  );
  const {planId} = useParams();

  const style = css`
    .loader-container {
      visibility: ${loading ? "visible" : "hidden"};
      position: fixed;
      margin-left: -75px;
      margin-bottom: 75px;
      left: 50%;
      bottom: 50%;
      width: 0;
      height: 0;
      z-index: 99;
    }
  `;

  useEffect(() => {

    async function fetchData(planId) {
      try {

        let created = "";
        let userId = 0;
        const server = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;
        let url = `http://${server}/plan/${planId}`;
        let obj = [];

        try {

          // get plan data
          const response = await fetch(url);
          if (response.ok) {
            // get data from the response
            obj = await response.json();
            userId = obj[0][0].studentId;
            created = obj[0][0].created;
            setCourses(obj);
            setPlanName(obj[0][0].planName);
            setUserId(obj[0][0].studentId);
            setStatus(parseInt(obj[0][0].status));
          } else {
            // we got a bad status code. send to 404 page
            props.history.push("/404");
            return;
          }

        } catch (err) {
          // send to 500 page if a server error happens while fetching plan
          console.log("An internal server error occurred. Please try again later.");
          props.history.push("/500");
          return;
        }

        // get active user information
        url = `http://${server}/user/${currentUserDev}`;
        let response = await fetch(url);
        if (response.ok) {
          // get data from the response
          obj = await response.json();
          setCurrentUser(
            {
              id: obj.userId,
              role: obj.role,
              firstName: obj.firstName,
              lastName: obj.lastName
            }
          );
        }

        // get plan user name
        url = `http://${server}/user/${userId}`;
        response = await fetch(url);
        if (response.ok) {
          // get data from the response
          obj = await response.json();
          setStudentName(obj.firstName + " " + obj.lastName);
          // add the intial review
          setActivity([{
            reviewId: 0,
            commentId: 0,
            status: 2,
            planId: planId,
            userId: userId,
            time: created,
            firstName: obj.firstName,
            lastName: obj.lastName
          }]);
        }

        // get plan activity
        url = `http://${server}/plan/${planId}/activity`;
        response = await fetch(url);
        setLoading(false);
        if (response.ok) {
          // get data from the response
          obj = await response.json();
          setActivity(prev => [...obj, ...prev]);
        }

      } catch (err) {
        // this is a server error
        console.log("An internal server error occurred. Please try again later.");
      }
    }
    fetchData(planId);

  }, [planId, props.history, currentUserDev]);

  async function handleAddComment(e) {
    setActivity(prev => [e, ...prev]);
  }

  async function handleChangeStatus(e) {
    setActivity(prev => [e, ...prev]);
    setStatus(parseInt(e.status));
  }

  return (
    <div className="view-plan" css={style}>
      <div className="loader-container">
        <BounceLoader
          size={150}
          color={"orange"}
        />
      </div>
      <NavBar showSearch={false} />
      <PlanMetadata studentName={studentName} userId={userId}
        planName={planName} status={status} currentUser={currentUser} />
      <PlanTable courses={courses} />
      {currentUser.role ? (
        <CreateReview currentUser={currentUser}
          onNewStatus={e => handleChangeStatus(e)} />
      ) : (
        null
      )}
      <ActivityFeed activity={activity} currentUser={currentUser}
        onNewComment={e => handleAddComment(e)} />
    </div>
  );
}
export default withRouter(ViewPlan);

ViewPlan.propTypes = {
  history: PropTypes.object
};