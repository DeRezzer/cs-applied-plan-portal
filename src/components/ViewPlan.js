import React from "react";
import PlanTable from "./PlanTable";
import PlanMetadata from "./PlanMetadata";
import PlanComments from "./PlanComments";

export default class ViewPlan extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      studentName: "",
      onid: 0,
      planName: "",
      status: -1,

      courses: [[], [{
        courseId: 0,
        credits: 0,
        courseName: "",
        courseCode: "",
        prerequisites: ""
      }], []]
    };

    this.searchPlans = this.searchPlans.bind(this);
  }

  async searchPlans() {

    const value = document.getElementById("search-plans-input").value;
    const url = `/api/plan/${value}`;
    let obj = [];

    try {
      const results = await fetch(url);
      obj = await results.json();
      this.setState({
        courses: obj
      });
      this.setState({
        studentName: "Luke Skywalker"
      });
      this.setState({
        planName: obj[0][0].planName
      });
      this.setState({
        onid: obj[0][0].studentId
      });
      this.setState({
        status: parseInt(obj[0][0].status)
      });
    } catch (err) {
      alert(err);
    }

  }

  render() {
    return (
      <div className="view-plan">

        <PlanMetadata studentName={this.state.studentName} onid={this.state.onid}
          planName={this.state.planName} status={this.state.status} />

        <PlanTable courses={this.state.courses} />

        <div className="search-plans-container">
          <form className="search-plans-form">
            <input id="search-plans-input" type="text" placeholder="Search.." name="search"/>
          </form>
          <button id="search-plans-button" type="submit" onClick={this.searchPlans}>Get Plan</button>
        </div>

        <PlanComments />

      </div>
    );
  }
}