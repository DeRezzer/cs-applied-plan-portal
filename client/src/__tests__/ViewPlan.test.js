/* eslint-disable no-undef */
import React from "react";
import {MemoryRouter} from "react-router-dom";
import renderer from "react-test-renderer";

import ViewPlan from "../components/view_plan/ViewPlan";

describe("View plan", () => {

  it("renders successfully", () => {
    const wrapper = renderer.create(
      <MemoryRouter initialEntries={["/viewPlan/10"]}>
        <ViewPlan history={{}}/>
      </MemoryRouter>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});