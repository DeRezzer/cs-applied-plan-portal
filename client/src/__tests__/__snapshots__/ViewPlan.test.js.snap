// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View plan renders successfully 1`] = `
<div
  className="css-6u0764-ViewPlan"
  id="view-plan-container"
>
  <div
    className="navbar-parent css-12g91iz-Navbar"
    id="navbar"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <p
        className="osu-logo"
      >
        OSU CS Applied Plan Portal
      </p>
    </a>
    <div
      className="right-container"
    >
      <div
        className="notification-dropdown css-3i4jcj-Notifications"
      >
        <button
          className="drop-button-notification"
          data-count={0}
        >
          Notifications
          <span
            className="badge"
          />
          <i
            className="fa fa-caret-down"
          />
        </button>
        <div
          className="dropdown-content"
        >
          <a
            href="/viewPlan/"
            onClick={[Function]}
          >
            <p>
              No new notifications
            </p>
          </a>
        </div>
      </div>
      <button
        className="logout-button css-1k1xu2y-Logout"
        onClick={[Function]}
      >
        Log Out
      </button>
    </div>
  </div>
  <div
    className="css-1rqemad-PlanMetadata"
    id="metadata-container"
  >
    <div
      className="plan-metadata"
    >
      <div
        className="metadata-field"
        id="plan-field"
      >
        <p
          className="field-type"
        >
          Plan Name:
        </p>
        <p
          className="field-text"
        >
          
        </p>
      </div>
      <div
        className="metadata-field"
        id="student-field"
      >
        <p
          className="field-type"
        >
          Student Name:
        </p>
        <p
          className="field-text"
        >
           
        </p>
      </div>
      <div
        className="metadata-field"
        id="email-field"
      >
        <p
          className="field-type"
        >
          Email:
        </p>
        <p
          className="field-text"
        >
          
        </p>
      </div>
      <div
        className="metadata-field"
        id="credit-field"
      >
        <p
          className="field-type"
        >
          Total Credits:
        </p>
        <p
          className="field-text"
        >
          0
        </p>
      </div>
      <div
        className="metadata-field"
        id="status-field"
      >
        <p
          className="field-type"
        >
          Plan Status:
        </p>
        <p
          className="field-text"
        >
          
        </p>
      </div>
      <div
        className="metadata-field button-field"
      >
        <button
          id="print-plan-button"
          onClick={[Function]}
        >
          Print Plan
        </button>
      </div>
    </div>
  </div>
  <div
    className="css-nk0226-PlanTable"
    id="table-container"
  >
    <table
      id="courses-table"
    >
      <thead>
        <tr>
          <th>
            Course
          </th>
          <th>
            Name
          </th>
          <th>
            Credits
          </th>
          <th
            className="restriction-header"
          >
            Restrictions
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <div
    className="css-exzccc-ActivityFeed"
    id="plan-activity"
  >
    <h2>
      Activity Feed
    </h2>
    <div
      className="css-174cigp-CreateComment"
      id="create-comment-container"
    >
      <button
        className="toggle-creation-button"
        onClick={[Function]}
      >
        Add Comment
      </button>
    </div>
  </div>
</div>
`;
